sql: add telemetry to cryptographic builtin functions

andyyang890 opened the pull request:
**sqltelemetry: add framework to count builtin function calls**

This patch adds a framework to collect telemetry for builtin function
calls. The counters will have a prefix of `sql.builtins` followed by
the builtin type and then the builtin name.

Release note: None

---

**sql: add telemetry to cryptographic builtin functions**

This patch adds telemetry to count the number of times any of the
cryptographic builtin functions are called.

Release note: None

---

Fixes #111323

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/111818)
<!-- Reviewable:end -->

andyyang890 committed:
sqltelemetry: add framework to count builtin function calls

This patch adds a framework to collect telemetry for builtin function
calls. The counters will have a prefix of `sql.builtins` followed by
the builtin type and then the builtin name.

Release note: None

andyyang890 committed:
sql: add telemetry to cryptographic builtin functions

This patch adds telemetry to count the number of times any of the
cryptographic builtin functions are called.

Release note: None

dikshant commented:
@rafiss if this was sent for every built in invocation would this slow down the built in even more?

rafiss reviewed (commented):
@dikshant i don't believe there is a perf impact, but i am a bit worried about how much volume the data could have in our telemetry storage.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/111818#-:-NgF1VaU6Dj_5t68uRYP:bxhs9np)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @abarganier and @andyyang890)

___
*[`pkg/sql/sem/builtins/pgcrypto_builtins.go` line 139 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/111818#-NgF1C9a5kdlzngX-hi4:-NgF1C9a5kdlzngX-hi5:b-8e2g7y) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f6f4a0d0d26b6c8b05e210a63c9bb4b5c7f11d4/pkg/sql/sem/builtins/pgcrypto_builtins.go#L139)):*
> ```Go
> 			ReturnType: tree.FixedReturnType(types.Bytes),
> 			Fn: func(_ context.Context, _ *eval.Context, args tree.Datums) (tree.Datum, error) {
> 				sqltelemetry.IncBuiltinFunctionCounter(sqltelemetry.CryptoBuiltinFunction, "digest")
> ```

i wonder if there's a way to add this call to telemetry so it happens at the place where `Fn` is invoked? or maybe so it happens during planing, so that we also can easily capture telemetry for builtins that are defined with a SQL body. doing it like that might make it so the category and name can automatically be inferred, instead of needing to duplicate it


<!-- Sent from Reviewable.io -->

andyyang890 commented:
Shelving for now, see https://github.com/cockroachdb/cockroach/issues/111323#issuecomment-1755988747.
